scrapscript.py

```diff
@@ -2630,6 +2869,149 @@ def test_serialize_record(self) -> None:
+        )
+
+    def test_serialize_apply(self) -> None:
+        obj = Apply(Var("f"), Var("x"))
+        self.assertEqual(
+            obj.serialize(),
+            {
+                b"func": {b"name": b"f", b"type": b"Var"},
+                b"arg": {b"name": b"x", b"type": b"Var"},
+                b"type": b"Apply",
+            },
+        )
+
+    def test_serialize_closure(self) -> None:
+        obj = Closure({"a": Int(123)}, Function(Var("x"), Var("x")))
+        self.assertEqual(
+            obj.serialize(),
+            {
+                b"env": {b"a": {b"type": b"Int", b"value": 123}},
+                b"func": {
+                    b"arg": {b"name": b"x", b"type": b"Var"},
+                    b"body": {b"name": b"x", b"type": b"Var"},
+                    b"type": b"Function",
+                },
+                b"type": b"Closure",
+            },
+        )
+
+
+class ObjectDeserializeTests(unittest.TestCase):
+    def test_deserialize_int(self) -> None:
+        msg = {"type": "Int", "value": 123}
+        self.assertEqual(Object.deserialize(msg), Int(123))
+
+    def test_deserialize_negative_int(self) -> None:
+        msg = {"type": "Int", "value": -123}
+        self.assertEqual(Object.deserialize(msg), Int(-123))
+
+    def test_deserialize_str(self) -> None:
+        msg = {"type": "String", "value": "abc"}
+        self.assertEqual(Object.deserialize(msg), String("abc"))
+
+    def test_deserialize_bytes(self) -> None:
+        msg = {"type": "Bytes", "value": b"abc"}
+        self.assertEqual(Object.deserialize(msg), Bytes(b"abc"))
+
+    def test_deserialize_var(self) -> None:
+        msg = {"type": "Var", "name": "abc"}
+        self.assertEqual(Object.deserialize(msg), Var("abc"))
+
+    def test_deserialize_bool_true(self) -> None:
+        msg = {"type": "Bool", "value": True}
+        self.assertEqual(Object.deserialize(msg), Bool(True))
+
+    def test_deserialize_bool_false(self) -> None:
+        msg = {"type": "Bool", "value": False}
+        self.assertEqual(Object.deserialize(msg), Bool(False))
+
+    def test_deserialize_bool_int_one(self) -> None:
+        msg = {"type": "Bool", "value": 1}
+        self.assertEqual(Object.deserialize(msg), Bool(True))
+
+    def test_deserialize_bool_int_zero(self) -> None:
+        msg = {"type": "Bool", "value": 0}
+        self.assertEqual(Object.deserialize(msg), Bool(False))
+
+    def test_deserialize_binary_add(self) -> None:
+        msg = {
+            "left": {"type": "Int", "value": 123},
+            "op": "ADD",
+            "right": {"type": "Int", "value": 456},
+            "type": "Binop",
+        }
+        obj = Binop(BinopKind.ADD, Int(123), Int(456))
+        self.assertEqual(Object.deserialize(msg), obj)
+
+    def test_deserialize_env_object(self) -> None:
+        obj = EnvObject({"x": Int(1)})
+        msg = {"env": {"x": {"type": "Int", "value": 1}}, "type": "EnvObject"}
+        self.assertEqual(Object.deserialize(msg), obj)
```

Comment:
I hadn't thought of deserializing the env object, but I like the idea the more I think of it. Great work